File: pokefeed/src/main/java/nl/mranderson/pokefeeds/video/VideoAdapter.java
package nl.mranderson.pokefeeds.video;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

import nl.mranderson.pokefeeds.R;
import nl.mranderson.pokefeeds.interfaces.ListItemListener;
import nl.mranderson.pokefeeds.video.model.Video;

public class VideoAdapter extends RecyclerView.Adapter<VideoAdapter.NewsItemViewHolder> {

    private final Context context;
    private final ListItemListener listener;
    private List<Video> items = new ArrayList<>();

    public VideoAdapter(Context context, ListItemListener listener) {
        this.context = context;
        this.listener = listener;
    }

    @Override
    public NewsItemViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new NewsItemViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.list_videoitem, parent, false));
    }

    @Override
    public void onBindViewHolder(NewsItemViewHolder holder, int position) {
        final Video item = items.get(position);
        holder.vTitle.setText(item.getTitle());
        Picasso.with(context)
                .load(item.getImageUrl())
                .fit()
                .centerCrop()
                .into(holder.vImage);
        holder.vImage.setOnClickListener(v -> listener.onItemTapped(item.getLink()));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public void update(List<Video> items) {
        this.items = items;
        notifyDataSetChanged();
    }

    public class NewsItemViewHolder extends RecyclerView.ViewHolder {
        TextView vTitle;
        ImageView vImage;

        NewsItemViewHolder(View v) {
            super(v);
            vTitle = v.findViewById(R.id.title_text);
            vImage = v.findViewById(R.id.item_image);
        }
    }
}

File: pokefeed/src/main/java/nl/mranderson/pokefeeds/video/VideoAdapter.kt
package nl.mranderson.pokefeeds.video

import android.content.Context
import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.squareup.picasso.Picasso
import nl.mranderson.pokefeeds.R
import nl.mranderson.pokefeeds.interfaces.ListItemListener
import nl.mranderson.pokefeeds.video.model.Video

class VideoAdapter(private val context: Context?, private val listener: ListItemListener) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var items = List<Video>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return VideoItemViewHolder(LayoutInflater.from(parent.context)
                .inflate(R.layout.list_videoitem, parent, false))
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        bindNews(holder, position)
    }

    private fun bindNews(holder: RecyclerView.ViewHolder, position: Int) {
        val item = items[position]
        if (holder is VideoItemViewHolder) {
            holder.title.text = item.title
            Picasso.with(context)
                    .load(item.imageUrl)
                    .fit()
                    .centerCrop()
                    .into(holder.image)
            holder.image.setOnClickListener { _ -> listener.onItemTapped(item.link) }
        }
    }

    override fun getItemCount(): Int {
        return items.size
    }

    fun update(newItems: List<Video>?) {
        if (newItems != null) {
            val diffResult = DiffUtil.calculateDiff(VideoDiffCallback(items, newItems))
            diffResult.dispatchUpdatesTo(this)
            items = newItems
        }
    }

    inner class VideoItemViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        var title: TextView = v.findViewById(R.id.title_text)
        var image: ImageView = v.findViewById(R.id.item_image)
    }
}
